Add support for shader includes.

glennw opened the pull request:
If the first line of a shader is //import followed by a comma
separated list of glsl filenames, prepend those as includes
for the shader. The idea here is that we'll start splitting
up prim_shared.glsl etc into smaller blocks that are faster
to parse and compile, and just generally more manageable.

This also supports having the VS and FS in the same glsl file. This
commit ports the debug_font and debug_color shaders to this style.

If we go ahead with this format, we'll incrementally port other
shaders to use this style, which requires only one file per shader.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/1588)
<!-- Reviewable:end -->

glennw commented:
r? @kvark and @nical 

It's always a bit of an open question how to manage shader includes and common code with GLSL. This seems to work reasonably well, and would allow us to modularize our shared GLSL code a lot. What do you think?

nical commented on webrender/src/device.rs:
```diff
@@ -46,10 +46,12 @@ const GL_FORMAT_BGRA_GL: gl::GLuint = gl::BGRA;
+const SHADER_KIND_VERTEX: &str = "#define WR_VERTEX_SHADER\n";
+const SHADER_KIND_FRAGMENT: &str = "#define WR_FRAGMENT_SHADER\n";
+const SHADER_IMPORT: &str = "//import ";
+const SHADER_LINE_MARKER: &str = "#line 1\n";
```
Neat!

kvark commented on webrender/src/device.rs:
```diff
@@ -770,23 +748,13 @@ impl Device {
-        let mut source = Vec::new();
-        source.extend_from_slice(s.as_bytes());
-        gl.shader_source(id, &[&source[..]]);
+        let sources: Vec<&[u8]> = sources.iter().map(|s| s.as_bytes()).collect();
```
I'd prefer us building a single source chunk per shader (as opposed to having an array of chunks), because OpenGL is the only API supporting multiple sources.

kvark commented on webrender/src/device.rs:
```diff
@@ -1190,128 +1158,137 @@ impl Device {
+        let mut shared_strings = Vec::new();
+        if let Some(shared_source) = get_shader_source(base_filename, &self.resource_override_path) {
+            parse_shader_source(shared_source,
+                                &self.resource_override_path,
```
time to learn the new formatting style ;)

glennw commented:
OK, updated to use ```#include```, support includes on any line, use a single source string, and also updated the shader validation test to be more representative of what we actually compile (and use the same shared code to build the shader string).

kvark commented:
Thank you!
@bors-servo r+

bors-servo commented:
:pushpin: Commit b19aabe has been approved by `kvark`

<!-- @bors-servo r=kvark b19aabec193cbcaf989b47edbfecac203e29a50a -->

bors-servo commented:
:hourglass: Testing commit b19aabec193cbcaf989b47edbfecac203e29a50a with merge 1b877000c45ce20a61c1a796046bb3704a7ba994...

bors-servo commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/servo/webrender/builds/266940253?utm_source=github_status&utm_medium=notification)
Approved by: kvark
Pushing 1b877000c45ce20a61c1a796046bb3704a7ba994 to master...
